atlasdb-api/src/main/java/com/palantir/atlasdb/keyvalue/api/KeyValueService.java

```diff
@@ -524,6 +524,14 @@ void deleteAllTimestamps(TableReference tableRef, Map<Cell, TimestampRangeDelete
     @Idempotent
     Set<TableReference> getAllTableNames();
 
+    /**
+     * Returns a list of up to {@code maxResults} tables stored in this key value service.
+     *
+     * This should be preferred over {@link #getAllTableNames()} if possible because the number of tables may be
+     * unbounded.
+     */
+    Iterable<TableReference> getLimitedTableNames(int maxResults);
```

Comment:
Should this be a paged result? What is the epoch for concurrent consistency (e.g. tables created while this is executing)? What happens if one specifies various values for `maxResults`, say `0`, `-1`, `Integer.MAX_VALUE`? I'm a little skeptical we even need these methods, but wondering if we actually do should instead consider something like `Stream<TableReference> getTableNames()` and let consumers filter/limit/transform/collect as desired.

Reply:
I was thinking about that, but it is significantly more complicated to implement correctly. Additionally the caller would need to call close() on the Stream, which is not something that is usually done in general (calling .collect() etc doesn't do it), so it would be really easy for people to leak resources if we made this a Stream.

CloseableIterator or something similar might make more sense, but it would also be more complicated to implement.

In general I think we should be discouraging anything that would try to enumerate all tables in the system. Clients can only interact with tables if they know about some Schema, and clients should be using the Schema to know what tables it needs to care about.